bionemo-recipes/recipes/esm2_native_te/dataset.py

```diff
@@ -215,7 +220,86 @@ def create_thd_dataloader(
         collate_fn=data_collator,
         num_workers=num_workers,
         pin_memory=True if not use_stateful_dataloader else False,
-        persistent_workers=True,
     )
-
     return train_dataloader, tokenized_dataset
+
+
+class CPAwareDataloader:
+    """A dataloader that is aware of context parallelism."""
+    def __init__(self, dataloader: StatefulDataLoader,
+                    cp_group: torch.distributed.ProcessGroup,
+                    cp_rank: int,
+                    ):
+        """Initialize the CPAwareDataloader."""
+        self.dataloader = dataloader
+        self.cp_rank = cp_rank
+        self.cp_group = cp_group
+        self.num_cp_ranks = cp_group.size()
+        self._iterator = None
+        
+
+    def __iter__(self):
+        """Make the dataloader iterable."""
+        self._iterator = iter(self.dataloader)
+        return self
+
+    def __next__(self):
+        """Get the batch from the dataloader for the current CP rank."""
+        batch = self._get_data_scatter_sharded()
+        batch['pad_between_seqs'] = True
+        return batch
+
+    def _get_data_scatter_sharded(self):
+        """
+        This function will get the batch from the dataloader on CP rank 0, and then determine 
+        the shards for all the different CP group members.
+        combined_batch = [<cp_rank_0_shard>, <cp_rank_1_shard>, ..., <cp_rank_n_shard>]
+        Then it will scatter the shards to the different CP group members.
+        The shards are then combined into a single batch and returned to the caller
+        for the current CP rank.
+
+        Scalability:
+            Rank 0's work grows linearly with CP size, but the other ranks do not need to store all the shards so they do not 
+            grow linearly with CP size.
+
+        Args:
+            None
+
+        Returns:
+            batch: The batch for the current CP rank.
+
+        """
+        if self.cp_rank == 0:
+            # Get data once, then make copies for each rank.
+            if self._iterator is None:
+                self._iterator = iter(self.dataloader)
+            batch = next(self._iterator)
+
+            # Now let's do the sharding on the batch for the different CP group members.
+            combined_batch = []
+            for cp_rank in range(self.num_cp_ranks):
+                input_ids_sharded, labels_sharded = get_batch_on_this_cp_rank(
+                    cu_seqlens_padded=batch["cu_seq_lens_q_padded"],
+                    input_ids_padded=batch["input_ids"],
+                    labels_padded=batch["labels"],
+                    cp_group=self.cp_group,
+                    qvk_format="thd",
+                    cp_rank=cp_rank,
+                )
+                batch_shard = dict(batch)
+                batch_shard["input_ids"] = input_ids_sharded
+                batch_shard["labels"] = labels_sharded
+                combined_batch.append(batch_shard)
+        else:
+            combined_batch = None
```

Comment:
we wanted to do this as a dataset.map call, right? otherwise this wont be done as part of the dataloader's prefetch